flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java

```diff
@@ -191,16 +200,29 @@ public ParallelismChange computeScaleTargetParallelism(
         double cappedTargetCapacity = averageTrueProcessingRate * scaleFactor;
         LOG.debug("Capped target processing capacity for {} is {}", vertex, cappedTargetCapacity);
 
-        int newParallelism =
+        Tuple2<Integer, Optional<String>> newParallelism =
                 scale(
```

Comment:
Instead of returning `Tuple2` here to return a message in addition to the new parallelism, can we pass the event handler to the scale method? We are already adding an additional parameter to the scale method. Eventually, we may want to reduce the number of parameters by grouping some of the parameters in wrapper objects, but no need to do this here.

Reply:
+1,
```suggestion
        int newParallelism =
            scale(
                vertex,
                currentParallelism,
                inputShipStrategies,
                (int) evaluatedMetrics.get(NUM_PARTITIONS).getCurrent(),
                (int) evaluatedMetrics.get(MAX_PARALLELISM).getCurrent(),
                scaleFactor,
                Math.min(currentParallelism, conf.getInteger(VERTEX_MIN_PARALLELISM)),
                Math.max(currentParallelism, conf.getInteger(VERTEX_MAX_PARALLELISM)),
                scalingLimitedMsg -> autoScalerEventHandler.handleEvent(
                    context,
                    AutoScalerEventHandler.Type.Warning,
                    SCALING_LIMITED,
                    scalingLimitedMsg,
                    SCALING_LIMITED + vertex + cappedTargetCapacity,
                    conf.get(SCALING_EVENT_INTERVAL)));
```
How about we pass in a function, This looks quite clear and does not require changing too many test cases.

Reply:
Resolved.